---
title: "Understanding the structure of Carbonic Anhydrase Using R"
author: "David Obe"
date: "2023-12-08"
output: html_document
---

## Carbonic Anhydrase

Carbonic anhydrase (CA EC 4.2.1.1) is an enzyme that is responsible for the reversible catalysis of carbon dioxide to bicarbonate and hydrogen ion. Although this reaction can proceed without an enzyme, the rate at which it proceeds naturally is too slow compared to the rate at which CO2 is produced within living system making CA important enzyme in regulation of pH homeostasis within the cell.

![*CA generalized reaction*](images/Screenshot%202023-11-20%20at%205.32.49%20PM.png){width="393"}

"CAs are widespread in both prokaryotes and eukaryotes, with eight genetic families encoding them, the α-, β-, γ-, δ-, ζ-, η-, θ-, and ι-classes (Supuran, 2023). The α-class is found in mammals and presently have 15 isoforms with different catalytic activity, subcellular localization and and tissue distribution such as those found in the cytosol, membrane-bonded CAs, mitochondrial, and secreted form, of all these isoforms, human CA II (hCA II) is the most researched (Supuran, 2008, Kim *et al.,* 2016).

CA has been linked to different physiological and pathological conditions and based on this, many modulators of CA has been developed to either inhibit or activate this enzyme in different disease conditions. Recently, CA II was suggested to be linked with Alzheimer disease a neurodegenerative disease that result into loss of memory.

![*Involvement of hCA various diseases and the isoforms responsible (Supuran 2023)*.](images/Screenshot%202023-11-20%20at%205.35.56%20PM.png){width="551"}

Therefore, understanding the structural changes that occur during the catalysis of this enzyme can help determine if this enzyme undergoes any motion during catalysis and how this is important for the activity of this enzyme. This analysis will also help in development of selective modulators of CAs in Alzheimer disease.

## Loading Packages

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



```{r, message=FALSE}
library(ggplot2)
library(gapminder)
library(dplyr)
library(gifski)
library(gganimate)
library(patchwork)
library(bio3d)
library(msa)
library(bio3d.view)
library(ggplot2)
library(ggrepel)
library(bigmemory)
library(rgl)
```

## Reading Data from Protein Data Bank (PDB) database.

PDB is a repository containing both experimentally resolved structures and computationally modeled structures. The PDB is under the watchful eyes of three musketeers forming a consortium (wwPDB). This consortium is between Europe (PDBe), Japan (PDBj), and the USA (RCSB PDB). They ensure that data published on these databases are free and accessible to the public. This search returned 237 hits.


```{r}
pdb <- read.pdb("1ca2")
```

```{r}
pdb
```

The major attributes of a pdb file can be accessed using

```{r}
attributes(pdb)
```

Since there are about 15 isoforms of CA, we tried to compare the structure of these isoforms to obtain their structural similarity, because structure mostly determine function.

```{r, message=FALSE, results="hide"}
ids <- c("2FOY_A", "1CA2", "4XIW_B","5IPZ_A", "1KEQ_A", "3FE4_A", "7P1A_A", "2W2J_A", "6RQU_A", "1JCZ_A", "3CZV_A", "1RJ5_A", "1DMX_A")
files <- get.pdb(ids, split = TRUE)
```

The get.pdb() function will download the requested files. Argument split = TRUE requests further that we want to extract particular chains, i.e. those specified by the _A suffix.

```{r, message=FALSE, results='hide'}
# Alignmnet
pdbs <- pdbaln(files, fit = TRUE, exefile="msa")
```
```{r, results='hide'}
pdbs
```

It's not enough to run alignment, we also would like to visualize our result. The active site of CA II contains His-94, His-96, Glu-106, His-119, and Thr-199. I would like to see the alignment in this positions. The result showed that this enzymes have different amino acids at these position which might sugest that they have different constituent at their active sites.

```{r}
pdbs$ali[1:13, c(94, 96, 106, 199)]
```

To further establish this, we processed to analyzing sequence similarity for these structures we also calculated the root mean standard deviation of this structures from each other.

```{r}
seqidentity(pdbs)
```

```{r, message=FALSE}
rd <- rmsd(pdbs)
```

We can also cluster these structures into trees to better visualize this similarity.

```{r}
hc <- hclust(as.dist(rd))
grps <- cutree(hc, k=3)

# plot results as dendogram
hclustplot(hc, k=3)
```
The figure above showed that 4XIW is different from all other structures. This might be because it is from *Chlamydomonas reinhartii* as shown below.

```{r}
anno <- pdb.annotate(ids)
print(unique(anno$source))
```

## B-factor

Carbonic anhdrase II is our focus therefore we proceeded to determine the dynamics of its (1CA2) structure. The B-factor also called Debye-Waller factor or temperature factor, is important in protein structure biology because it signifies the displacement of atoms from thier mean position (Pandey *et al.,* 2023). The B-factor also give insight into flexibility, solvent accessibility, thermal stability, and activity of the region in protein.

```{r}
plot.bio3d(pdb$atom$b[pdb$calpha], sse=pdb, type="l", ylab="B-factor")
```
## Retrieving sequence data from PDB

From the graph above, the end terminals of this protein seems to be more dynamic than the core part of the protein. To establish this, we decided to look into the possible conformation that CA II can possibly occur.

```{r}
pdbs <- get.seq("1ca2")
blast <- blast.pdb(pdbs)
```

1CA2 is a monomeric metalloprotein containing zinc (Avvaru *et al.,* 2009). This structure is made up 256 amino acid residues and it was obtained from the erythrocytic of human (Eriksson *et al.,* 1988).

![Figure 3: Active site residues of CA (https://academic-accelerator.com/encyclopedia/carbonic-anhydrase)](images/Active_site_of_CAII_(cropped).jpg){width="252"}

The visualization of our blast search showed that their is a major shift in E-value scores which suggest that some of the sequences obtained are not relevant to our research and might result into false positives. Therefore, we set a cutoff value for our analysis. This yielded a total of 166 hits that have similar E-values as shown in the graph below.

```{r}
plot.blast(blast)
```

Based on the E-value and identity score, we set a cutoff for our research. This will reduce the unrelated sequences and decrease processing time.

```{r}
hits <- plot(blast, cutoff = 500)
```

```{r}
hits$pdb.id
```
It is the structure that we need to work with, therefore we use the get.pdb() to retrieve the structures of our 166 closely related structures.

```{r}
# Retrieve data from PDB
files <- get.pdb(hits$pdb.id, path = "pdbs", split = TRUE, gzip = TRUE)
```
The sequences obtained are then aligned using the pdbaln() function.

```{r, message=FALSE, results='hide'}
# Align structure files
pdbs <- pdbaln(files, fit=TRUE, exefile="msa")
```
We visualized the alignment to identify gaps in our sequence alignment.

```{r}
# Visualize alignment
ids <- basename.pdb(pdbs$id)
plot(pdbs, labels=ids)
```
To identify conserved domains in our data we used the conserv() function.

```{r}
# calculate conserved domain
cons <- conserv(pdbs, method = "entropy22")
```

```{r}
# SSE annotation
sse <- pdbs2sse(pdbs, ind = 1, rm.gaps = FALSE)
```

```{r}
# plot conservation per residue 
plotb3(cons, sse = sse, ylab = "sequence entropy")
```
The plot showed the conserved domains with the black (alpha helix) and gray bar (beta sheets). To know why we have such a highly conserved regions in our structures, we check the annotation/source of our sequences.

```{r}
anno <- pdb.annotate(ids)
print(unique(anno$source))
```

## Data cleaning

The alignment above shows that we have gap in some sequences and some might be redundant sequences, this can negatively impact our result and slow our processing time down. Therefore, need to remove the gaps and the sequences that are redundant.

```{r}
# remove structure with missing residues
conn <- inspect.connectivity(pdbs, cut = 4.05)
trim <- trim.pdbs(pdbs, row.inds=which(conn))

# which structures were omitted?
which(!conn)
```

```{r, message=FALSE, results='hide'}
# to remove conformational redundant structures
rd <- filter.rmsd(trim$xyz, cutoff = 0.1, fit = TRUE)
trim <- trim.pdbs(trim, row.inds = rd$ind)
head(trim)
trim$id
```


```{r}
# list of PDB codes of our final selection

ids <- unlist(strsplit(basename(trim$id), split=".pdb"))
```

```{r}
# Visualize my pdb data
print(trim, alignment = FALSE)
```

## Principal component analysis
PCA is a statistical method used in transforming data sets to few important components grouping them based on similarity or variation with an ensemble of structures. For good analysis we first have to superimpose all the structures.

```{r, message=FALSE, results='hide'}
# locate invariant core
core <- core.find(trim)
```


```{r}
# Now that we have determined the core, we can superimpose all structures to core
trim$xyz = pdbfit(trim, core$c0.5A.xyz)
pc.xray <- pca(trim)
```

```{r}
# Now that we have determined the core, we can superimpose all structures to core
trim$xyz = pdbfit(trim, core$c0.5A.xyz)
```

```{r}
# Identifying and removing gaps
gap.pos <- gap.inspect(trim$xyz)
gap.res <- gap.inspect(trim$ali)
```

```{r}
# Perform PCA
pc.xray <- pca.xyz(trim$xyz[,gap.pos$f.inds])
plot(pc.xray)
```

We need to calculate the pairwise root mean standard deviation (RMSD) for structural ensemble. 

```{r, message=FALSE}
# calculate RMSD
rd <- rmsd(trim)
```


```{r}
# structure based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=4)
plot(pc.xray$z[,1:2], col="gray50", pch=16, cex=1.3,
ylab="Pricipal Component 2", xlab="Principal Component 1")
points(pc.xray$z[,1:2], col=grps.rd, pch=16, cex=0.9)
```

```{r}
plot(pc.xray$z[,2:3], col="gray50", pch=16, cex=1.3,
     ylab="Pricipal Component 2", xlab="Principal Component 3")
points(pc.xray$z[,2:3], col=grps.rd, pch=16, cex=0.9)
```

```{r, echo=FALSE}
# To visualize use
pc1 <- mktrj(pc.xray, pc=1, file="pc_1.pdb")
```

## Adding lebel to the PCA plot

```{r}
# Labeling IDs to the plot for PC 1 and 2
df <- data.frame(x=pc.xray$z[,1], y=pc.xray$z[,2])
col <- as.factor(grps.rd)
```

```{r, message=FALSE}
# plot of PC2 against PC1
p <- ggplot(df, aes(x,y))+
geom_point(aes(col=col), size=2) +
xlab("Principal Component 1") +
ylab("Principal Component 2") +
scale_color_discrete(name="Clusters") +
geom_text_repel(aes(label=ids), max.overlaps = 20) + theme(legend.position = "none")
p
```

```{r}
# Labeling IDs to the plot for PC 1 and 2
df <- data.frame(x=pc.xray$z[,2], y=pc.xray$z[,3])
col <- as.factor(grps.rd)
```

```{r}
# plot of PC2 against PC1
q <- ggplot(df, aes(x,y))+
geom_point(aes(col=col), size=2) +
xlab("Principal Component 2") +
ylab("Principal Component 3") +
scale_color_discrete(name="Clusters") +
geom_text_repel(aes(label=ids), max.overlaps = 20) +
  theme(legend.position = "none")
q
```


## Fluctuation analysis

Proteins exist can exist in open (black), closed (red), or occluded (green). This analysis seeks to understand the state in which CA II exist.

```{r, echo=FALSE}
mktrj(pc.xray, pc=1,
resno=trim$resno[1, gap.res$f.inds],
resid=trim$resid[1, gap.res$f.inds])
```

```{r, message=FALSE, results='hide'}
# clustering by conformation
pdbfit(trim.pdbs(trim, row.inds = which(grps.rd==1)), outpath="grps1") ## closed
modes <- nma(trim, rm.gaps=TRUE, ncore=4, bigmem=TRUE)

#print modes
print(modes)
```

```{r}
# Visualize result
plot(modes, pdbs=trim, col=grps.rd, label=NULL)
```

The result which is also in agreement with the PCA showed that closed (red) and open (black) conformations are the major states that CA II exist in. To better enhance our understanding, we clustered the conformation using dendogram and compared the two main conforamtions.

```{r}
# clustering
hc.nma <- hclust(as.dist(1-modes$rmsip))
grps.nma <- cutree(hc.nma, k=4)
heatmap(1-modes$rmsip, distfun = as.dist, labRow = ids, labCol = ids,
ColSideColors = as.character(grps.nma), RowSideColors = as.character(grps.nma))
cols <- grps.rd
cols[which(cols!=1 & cols!=2)]=NA
```
```{r}
plot(modes, pdbs=trim, col=cols, signif=TRUE)
```


## Measure for modes comparison
We can cluster structures based on fluctuation using squared inner product (SIP), Root mean squared inner product (RMSIP), covariance overlap, Bhattacharyya coefficient, and PCA of covariance matrices.

```{r}
# Similarity of atomic fluctuations
sip <- sip(modes)
hc.sip <- hclust(as.dist(1-sip), method = "ward.D2")
grps.sip <- cutree(hc.sip, k=4)
```

```{r}
# Squared inner product (SIP) plot
hclustplot(hc.sip, k=4, colors = grps.rd, labels = ids, cex=0.7, main="SIP", fillbox = FALSE)
par(fig=c(.65, 1, .45, 1), new = TRUE)
plot(pc.xray$z[,1:2], col="grey50", pch=16, cex=1.1,
ylab="", xlab="", axes=FALSE, bg="red")
points(pc.xray$z[,1:2], col=grps.sip, pch=16, cex=0.9)
box()
```

```{r}
# Root mean squared inner product (RMSIP)
rmsip <- rmsip(modes)
hc.rmsip <- hclust(dist(1-rmsip), method = "ward.D2")
grps.rmsip <- cutree(hc.rmsip, k=4)
hclustplot(hc.rmsip, k=4, colors = grps.rd, labels = ids, cex=0.7, main="RMSIP", fillbox = FALSE)
par(fig=c(.65, 1, .45, 1), new = TRUE)
plot(pc.xray$z[,1:2], col="grey50", pch=16, cex=1.1,
ylab="", xlab="", axes=FALSE, bg="red")
points(pc.xray$z[,1:2], col=grps.sip, pch=16, cex=0.9)
box()
```

```{r}
# covariance overlap
co <- covsoverlap(modes, subset=200)
hc.co <- hclust(as.dist(1-co), method = "ward.D2")
grps.co <- cutree(hc.co, k=4)
hclustplot(hc.co, k=4, colors = grps.rd, labels = ids, cex=0.7, main="covariance overlap", fillbox = FALSE)
par(fig=c(.65, 1, .45, 1), new = TRUE)
plot(pc.xray$z[,1:2], col="grey50", pch=16, cex=1.1,
ylab="", xlab="", axes=FALSE, bg="red")
points(pc.xray$z[,1:2], col=grps.sip, pch=16, cex=0.9)
box()

```

```{r}
# Bhattacharyya coefficient
covs <- cov.enma(modes)
bc <- bhattacharyya(modes, covs=covs)
hc.bc <- hclust(dist(1-bc), method = "ward.D2")
grps.bc <- cutree(hc.bc, k=4)
```

```{r}
hclustplot(hc.bc, k=4, colors = grps.rd, labels = ids, cex=0.7, main="Bhattacharyya coefficient", fillbox = FALSE)
par(fig=c(.65, 1, .45, 1), new = TRUE)
plot(pc.xray$z[,1:2], col="grey50", pch=16, cex=1.1,
ylab="", xlab="", axes=FALSE, bg="red")
points(pc.xray$z[,1:2], col=grps.sip, pch=16, cex=0.9)
box()
```

```{r}
# PCA of covariance matrices
pc.covs <- pca.array(covs)
hc.covs <- hclust(dist(pc.covs$z[,1:2]), method = "ward.D2")
grps.covs <- cutree(hc.covs, k=4)
```
```{r}
hclustplot(hc.covs, k=4, colors = grps.rd, labels = ids, cex=0.7, main="PCA of covariance matrices", fillbox = FALSE)
par(fig=c(.65, 1, .45, 1), new = TRUE)
plot(pc.xray$z[,1:2], col="grey50", pch=16, cex=1.1,
ylab="", xlab="", axes=FALSE, bg="red")
points(pc.xray$z[,1:2], col=grps.sip, pch=16, cex=0.9)
box()
```

## Conclusion

My goal for registering for this class was achieved due to patient teaching and prompt feedback from Dr. Shizuka. I am glad to find an R package created by Dr. Barry Grant that relates to my field of study - Structural Biology. Analysis was performed using Bio3D a product of grant lab.


## References
Grant B.J, Rodrigues A.P.C., ElSawy K.M., McCammon J.A., Caves L.S.D. (2006). Bio3d: an R package for the comparative analysis of protein structures, Bioinformatics, Volume 22, Issue 21, November 2006, Pages 2695–2696, https://doi.org/10.1093/bioinformatics/btl461

Supuran, C. Carbonic anhydrases: novel therapeutic applications for inhibitors and activators. Nat Rev Drug Discov 7, 168–181 (2008). https://doi.org/10.1038/nrd2467

Supuran CT. Carbonic anhydrase versatility: from pH regulation to CO2 sensing and metabolism. Front Mol Biosci. 2023 Nov 10;10:1326633. doi: 10.3389/fmolb.2023.1326633. PMID: 38028557; PMCID: PMC10676200.